Feature/bsk-1043: Active set allocation and bug fix in forceTorqueThrForceMapping

E-Krantz opened the pull request:
* **Tickets addressed:** #1043 
* **Review:** By commit
* **Merge strategy:** Merge (no squash)

## Description

This PR addresses two issues in the `forceTorqueThrForceMapping` module:

1. **Bug fix:**
   Corrects the check for zero rows in the DG matrix, which previously used a hardcoded value instead of the actual number of thrusters. This ensures thrusters are zeroed out correctly based on the true system configuration.

2. Small code clean-up 

All changes are made in logically organised commits.

## Verification

* Updated the unit tests for `forceTorqueThrForceMapping` to compare the actual net force and torque produced by the allocated thruster forces, rather than checking for specific per-thruster values (since multiple valid allocations can exist).
* Introduced a new helper, `calculate_force_torque()`, to compute the resultant force and torque from the set of thruster outputs using standard physics equations.
* The tests now validate that the allocated solution produces the correct net force and torque (within tolerance), using `unitTestSupport.compareArray()`.
* All tests pass after these updates.

## Documentation

* Inline code comments have been updated for clarity.

## Future work

* Update public documentation to fully reflect the new algorithm and limitations.
* Additional unit tests could be added for edge cases.

E-Krantz commented:
Fixed Windows build failures caused by Variable Length Arrays not being supported in MSVC.

Issue with build on macOS seems unrelated to this PR. `requirements_doc.txt` has a syntax error:
`sphinx>7.0.0,=<8.2.3`
which should be replaced with
`sphinx>7.0.0,<=8.2.3`

Should I add this fix to this PR?

schaubh commented:
Mm, curious how this MSVC issue didn't show up before?  

Regarding the requirements syntax error, good eye, yes, please include it in this PR.  I hope to review this PR before the end of the week.

juan-g-bonilla commented:
> Mm, curious how this MSVC issue didn't show up before? @juan-g-bonilla , have you seen this MSVC limitation before?
> 
> Regarding the requirements syntax error, good eye, yes, please include it in this PR. I hope to review this PR before the end of the week.

Never seen it. Support for VLAs is not in the C++ standard and has been optional in C since C11, so it's good to get rid of it if we can. There are compiler/compiler extensions that allow support for VLAs, so if we happened to have been using those it's not unreasonable that this slipped through.

Thanks @E-Krantz for the PR!

E-Krantz commented:
Hi. Thanks for the review. Happy to contribute to a great simulator.

Two new commits:
- Fixed the syntax error in `requirements_docs.txt`.
- Updated documentation rst for `forceTorqueThrForceMapping` module to match the new algorithm.

It is no issue to work with fixed-size-arrays, just something that didn't show up when building on Linux. It was an easy fix.

schaubh commented on src/fswAlgorithms/effectorInterfaces/forceTorqueThrForceMapping/forceTorqueThrForceMapping.rst:
```diff
@@ -90,7 +90,7 @@ The total force and torque on the spacecraft may be represented as:
     \end{bmatrix} = [D]\mathbf{F}
 
-The force required by each thruster is first computed by the following minimum-norm solution. Any rows within the :math:[D] matrix
+The force required by each thruster is first computed by the following minimum-norm solution. Any rows within the :math:`[D]` matrix
```
good catch!

E-Krantz commented on src/fswAlgorithms/effectorInterfaces/forceTorqueThrForceMapping/forceTorqueThrForceMapping.c:
```diff
@@ -93,10 +92,14 @@ void Update_forceTorqueThrForceMapping(forceTorqueThrForceMappingConfig *configD
     CmdTorqueBodyMsgPayload cmdTorqueInMsgBuffer;  //!< local copy of message buffer
     CmdForceBodyMsgPayload cmdForceInMsgBuffer;  //!< local copy of message buffer
     THRArrayCmdForceMsgPayload thrForceCmdOutMsgBuffer;  //!< local copy of message buffer
+    THRArrayConfigMsgPayload thrConfigInMsgBuffer;  //!< local copy of message buffer
```
Agreed. I will remove the need for thrConfigInMsgBuffer.

E-Krantz commented on src/fswAlgorithms/effectorInterfaces/forceTorqueThrForceMapping/forceTorqueThrForceMapping.c:
```diff
@@ -93,10 +92,14 @@ void Update_forceTorqueThrForceMapping(forceTorqueThrForceMappingConfig *configD
+    /* always zero the output message buffers before assigning values */
     thrForceCmdOutMsgBuffer = THRArrayCmdForceMsg_C_zeroMsgPayload();
 
+    /* Read thruster configuration once for the entire function */
```
Understood. I will remove it.

E-Krantz commented:
Thank you for the detailed feedback! My intention with this PR was primarily to address what appeared to me as a physical inconsistency in the existing algorithm, rather than to fundamentally alter its behaviour. However, given your comments, I'm happy to revert this PR to a straightforward bug fix.

A few clarifications/questions:

* You mentioned:

  > "The current approach is specific to a THR configuration that is configured to produce pure torques."
  
  From the current documentation, I understood this module as being intended to produce both requested torques and forces. Is this understanding correct? Is the primary design assumption that most use cases are pure-torque configurations?

* Regarding:

  > "Thus, it has a null space where all thrusters can fire at the same time and not produce a force or torque."
  
  Is this considered desirable? I initially viewed this scenario as non-ideal, hence my attempt to modify the behaviour.

* Regarding computational cost:

  You're correct; the iterative approach performs multiple least-squares solutions (up to `numThrusters` iterations). In practice, this is still relatively efficient for typical thruster arrays, though clearly slower than the current single-computation solution.

* Regarding your question:

  > "Does it revert to the same answer as the original algorithm? Does it act better if the THR configuration is not a pure-torque setup?"
  
  The current algorithm subtracts the minimum computed thrust from all thrusters, affecting thrusters that might be orthogonal or nearly orthogonal to the minimum-thrust thruster. This sometimes leads to final force/torque outputs deviating from the requested commands. Thus, the iterative algorithm doesn't produce identical results; instead, it generally produces allocations that more closely match the requested net forces and torques.

Currently, this module isn't utilised by any example scenarios or other modules within Basilisk, and thus the differences haven’t yet affected existing scenarios in the repo. I encountered unexpected behaviour, specifically when trying to use it for combined force and torque allocation, beyond the previously mentioned thruster-zeroing bug.

Would you like me to revert the documentation and algorithm changes, keeping this PR strictly as a bug fix?
If there’s future interest, the iterative method could potentially be explored separately.

schaubh commented:
Yes, I'd like to keep this as a bug fix for now.  If other THR force mapping algorithms are to be used, these should be put into separate modules.  As always, each has their benefits and drawbacks.

Regarding this algorithm, it does produce both desired force and torques.  But, it does assume an attitude control thruster configuration that produces pure torques.  Mathematically this math assumes that if all thrusters are firing by the same amount, no net force or torque is produced.  I should update the documentation to make this clearer.

Your algorithm might be more general and give better results if this above condition is not met.  We are looking at general optimizers for our work currently to yield these constrained thruster solutions.

E-Krantz commented:
Understood. I will revert this to a simple bug fix then.

Would you like to keep the change to the unit test, i.e., compare with the requested torque/force instead of relying on precalculated values?

E-Krantz commented:
Reverted to a bug-fix and updated unit test for `forceTorqueThrForceMapping`

- Change to `forceTorqueThrForceMapping.c` is now only a bug-fix to resolve the issue where sometimes the wrong rows are zeroed out.
- Unit test for `forceTorqueThrForceMapping` now compares resultant torque/force with requested torque/force instead of relying on hardcoded values.

I noticed that the previous unit test used `unitTestSupport.compareArray()` instead of `unitTestSupport.compareVector()`. compareArray appears to only compare the first 3 elements in the arrays, whereas compareVector() compares all elements.

schaubh commented:
thanks for the edits and support.  Need to get through a few more papers to edits, expect to return to this by end of week.

schaubh commented on requirements_doc.txt:
```diff
@@ -1,6 +1,6 @@
 breathe==4.36.0
 docutils==0.21.2
 recommonmark==0.7.1
-sphinx>7.0.0,=<8.2.3
```
Please break up this commit into three.  One should have this requirements fix.  The 2nd should have the *.c file changes.  The final one should have the updated test file.

Also, your commit message has a first line that is too long.  Note on GitHub how the first line is shown as a title for this commit.  With a long first line the bold title is truncated oddly.  Just have short first lines, and add additional info in the 2nd line in the commit.

schaubh reviewed (changes requested):
Nice PR.  Thanks for the support.  Almost there. Request a simple clean up of the commits and commit message.  Further, this branch is missing updated release notes and known issues.  Please update

/docs/source/Support/bskReleaseNotes.rst

where you add a line saying:
```
The module :ref:`forceTorqueThrForceMapping` correct a false behavior in ``Reset()`` where not all thruster settings were zero'd on reset.
```

Next, please add information to 

/docs/source/Support/bskKnownIssues.rst

Saying:

```
The ``Reset()`` function in :ref:`forceTorqueThrForceMapping` was not working properly. This has been addressed in the current release.
```

schaubh commented:
I updated the PR text to reflect the reduced set of changes in this branch.

schaubh commented:
Howdy, just checking in on this PR.  I know my requested edits are very small ones.  If you prefer, if you give me write access to your branch I can do these myself.  Thanks again for your contribution and sharing the issue you found.

E-Krantz commented:
Howdy. I've corrected the PR following your comments.

schaubh commented:
Thanks. Can you rebase on latest develop? I can then review and close this PR.

E-Krantz commented:
Certainly!

schaubh reviewed (approved):
Thanks for the contribution!

E-Krantz commented:
I'm happy to contribute. Thank you! :rocket:
